go/enclave/l2chain/l2_chain.go

```diff
@@ -893,6 +902,53 @@ func (lc *L2Chain) checkAndStoreBatch(batch *core.Batch) error {
 	return nil
 }
 
+// rollbackL2ToLatestValidBatch is called when the currHead's L1 proof block is no longer on the canonical fork
+// it will scan back through batch chain until it finds a batch associated with the canonical L1 chain
+func (lc *L2Chain) rollbackL2ToLatestValidBatch(newL1Head *types.Block) error {
+	currHead, err := lc.storage.FetchHeadBatch()
+	if err != nil {
+		return fmt.Errorf("l2 rollback after fork: unable to fetch head batch - %w", err)
+	}
+	latestValidBatch := currHead
+	// check if batch is tied to the new L1 head or an ancestor of it
+	for !lc.isBatchLinkedToCanonicalL1Block(latestValidBatch, newL1Head) {
+		if latestValidBatch.Header.Number.Cmp(big.NewInt(1)) <= 0 {
+			// reached genesis without finding a canonical L1 block, something has gone critically wrong
+			lc.logger.Crit("reached genesis batch without finding a batch linked to canonical L1 block, aborting...")
+		}
+		// if not then move to the parent batch to check if that is L1-canonical
+		latestValidBatch, err = lc.storage.FetchBatch(latestValidBatch.Header.ParentHash)
+		if err != nil {
+			return fmt.Errorf("l2 rollback after fork: could not find parent batch - %w", err)
+		}
+	}
+	if latestValidBatch.Hash() != currHead.Hash() {
+		err := lc.storage.SetHeadBatchPointer(latestValidBatch)
+		lc.logger.Info("L2 chain head has been rolled back",
```

Comment:
the log doesn't appear correct. 
there is a dingling "latestValidBatch.Hash()"

Reply:
Good spot, thanks!